### PR TITLE
feat(ctb): `OutputBisection` actor tests

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Introduces a `GameSolver` harness for actor tests against the `OutputBisectionGame` contract. Each actor is an abstraction
around the game solving logic, capable of participating in an `OutputBisectionGame` backed by the `AlphabetVM2`.

### Test Cases

> **Note**
> We don't have to cover all of these, though it would be nice. `op-e2e` gives us more concrete coverage, these should primarily help drive contracts changes w/o the weight of `op-e2e`.

- [x] [Fuzz] Correct root, same trace length between actors - `DEFENDER_WINS`
- [x] Correct root, defense step by dishonest actor - `DEFENDER_WINS`.
- [x] Incorrect root, defense step by honest actor - `CHALLENGER_WINS`.
- [x] Correct root, attack step by dishonest actor from the absolute prestate - `DEFENDER_WINS`
- [x] Incorrect root, attack step by honest actor from the absolute prestate - `CHALLENGER_WINS`
- [x] Correct root, attack step in secondary execution trace subgame by dishonest actor - `DEFENDER_WINS`
- [x] Incorrect root, attack step in secondary execution trace subgame by honest actor - `CHALLENGER_WINS`
- [x] Correct root, defense step in secondary execution trace subgame at final insn by dishonest actor - `DEFENDER_WINS`
- [x] Incorrect root, defense step in secondary execution trace subgame at final insn by honest actor - `CHALLENGER_WINS`
- [ ] Variable trace lengths
- [ ] Variable `l2Outputs` lengths (?)

**Metadata**
closes https://github.com/ethereum-optimism/client-pod/issues/255
closes https://github.com/ethereum-optimism/client-pod/issues/253

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #8537** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8537?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #8538** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8538?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #8536** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8536?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8536?utm_source=stack-comment).

File: packages/contracts-bedrock/test/mocks/AlphabetVM2.sol [hunk 1]
```diff
@@ -23,23 +24,27 @@ contract AlphabetVM2 is IBigStepper {
...
         returns (bytes32 postState_)
     {
         uint256 traceIndex;
         uint256 claim;
         if ((keccak256(_stateData) << 8) == (Claim.unwrap(ABSOLUTE_PRESTATE) << 8)) {
             // If the state data is empty, then the absolute prestate is the claim.
-            traceIndex = 0;
-            (claim) = abi.decode(_stateData, (uint256));
-            claim = claim + uint256(oracle.loadLocalData(4, _localContext, 0, 32, 0));
+            (bytes32 dat,) = oracle.readPreimage(PreimageKeyLib.localizeIdent(4, _localContext), 0);
```
Comment (Inphi) [inline packages/contracts-bedrock/test/mocks/AlphabetVM2.sol]: nit: should use `LocalPreimageKey.STARTING_L2_BLOCK_NUMBER`, even though this is just a stub identity.

File: packages/contracts-bedrock/test/actors/OutputBisectionActors.sol [hunk 2]
```diff
@@ -286,56 +282,121 @@ contract HonestGameSolver is GameSolver {
...
+            GameSolver.Move memory localMove = moves[i];
+
+            // If the move is a step, we first need to add the starting L2 block number to the `PreimageOracle`
+            // via the `OutputBisectionGame` contract.
+            // TODO: This is leaky. Could be another move kind.
+            if (localMove.kind == GameSolver.MoveKind.Step) {
+                bytes memory moveData = localMove.data;
+                uint256 challengeIndex;
+                assembly {
+                    challengeIndex := mload(add(moveData, 0x24))
```
Comment (Inphi) [inline packages/contracts-bedrock/test/actors/OutputBisectionActors.sol:387]: Is this meant to be a bitshift? What's the relationship between the step data and `challengeIndex`?

File: packages/contracts-bedrock/test/dispute/OutputBisectionGame.t.sol [hunk 3]
```diff
@@ -632,11 +641,549 @@ contract OutputBisectionGame_Test is OutputBisectionGame_Init {
...
+        vm.label(address(honest), "HonestActor");
+        vm.label(address(dishonest), "DishonestActor");
+    }
+
+    /// @dev Helper to exhaust all moves from both actors.
+    function _exhaustMoves() internal {
+        while (true) {
+            // Allow the dishonest actor to make their moves, and then the honest actor.
+            (uint256 numMovesA,) = dishonest.move();
+            (uint256 numMovesB,) = honest.move();
```
Comment (Inphi) [inline packages/contracts-bedrock/test/dispute/OutputBisectionGame.t.sol]: in a 1v1, we expect every honest move to be successful. Perhaps it's worth asserting this here?

File: packages/contracts-bedrock/test/actors/OutputBisectionActors.sol (hunk 2 above)
Comment (clabby) [reply packages/contracts-bedrock/test/actors/OutputBisectionActors.sol:387]: Nope; This is correct. This is a bit hacky, but we can always pull out the 32 byte `challengeIndex` due to the standard encoding.

The `step` calldata signature is: `step(uint256 _claimIndex, bool _isAttack, bytes calldata _stateData, bytes calldata _proof)`

The way that this ends up being abi-encoded is such that `_claimIndex` (or `challengeIndex` here) is always directly after the selector.

Comment (Inphi) [reply packages/contracts-bedrock/test/actors/OutputBisectionActors.sol:387]: Got it. (I need to sleep more)

**Review by Inphi — APPROVED** [review]
Very nice!

**Pull request merged**